File: ts/src/samchon/protocol/distributed/DistributedSystemArrayMediator.ts
/// <reference path="../../API.ts" />

/// <reference path="DistributedSystemArray".ts" />

namespace samchon.protocol.distributed
{
	export abstract class DistributedSystemArrayMediator
		extends DistributedSystemArray
	{
		/**
		 * @hidden
		 */
		private mediator_: parallel.MediatorSystem;

		/* ---------------------------------------------------------
			CONSTRUCTORS
		--------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();

			this.mediator_ = null;
		}

		protected abstract createMediator(): parallel.MediatorSystem;

		protected startMediator(): void
		{
			if (this.mediator_ != null)
				return;

			this.mediator_ = this.createMediator();
			this.mediator_.start();
		}

		/* ---------------------------------------------------------
			ACCESSOR
		--------------------------------------------------------- */
		public getMediator(): parallel.MediatorSystem
		{
			return this.mediator_;
		}

		/* ---------------------------------------------------------
			MESSAGE CHAIN
		--------------------------------------------------------- */
		/**
		 * @hidden
		 */
		protected _Complete_history(history: parallel.PRInvokeHistory): boolean
		{
			let ret: boolean = super._Complete_history(history);
			if (ret == true)
				this.mediator_["complete_history"](history.getUID());

			return ret;
		}
	}
}

File: cpp/samchon/protocol/distributed/DistributedSystemArrayMediator.cpp
#include <samchon/protocol/distributed/DistributedSystemArrayMediator.hpp>

#include <samchon/protocol/InvokeHistory.hpp>
#include <samchon/protocol/parallel/MediatorSystem.hpp>

using namespace std;
using namespace samchon::protocol;
using namespace samchon::protocol::distributed;
using namespace samchon::protocol::parallel;

/* ---------------------------------------------------------
	CONSTRUCTORS
--------------------------------------------------------- */
DistributedSystemArrayMediator::DistributedSystemArrayMediator()
	: DistributedSystemArray(),
	ParallelSystemArrayMediator()
{
}

DistributedSystemArrayMediator::~DistributedSystemArrayMediator()
{
}

/* ---------------------------------------------------------
	HISTORY HANDLER
--------------------------------------------------------- */
auto DistributedSystemArrayMediator::_Complete_history(shared_ptr<InvokeHistory> history) -> bool
{
	bool ret = DistributedSystemArray::_Complete_history(history);
	if (ret == true)
		getMediator()->_Complete_history(history->getUID());

	return ret;
}